### PR TITLE
Add publishing of dependency constraints to Gradle module metadata

### PR DESCRIPTION
**Pull request by jjohannes**
See #3551. Follow up to #3588.

### CODE REVIEW

File: subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishJavaIntegTest.groovy [hunk 1]
```diff
@@ -201,6 +201,67 @@ class MavenPublishJavaIntegTest extends AbstractMavenPublishIntegTest {
...
+        given:
+        createBuildScripts("""
+
+            ${jcenterRepository()}
+
+            dependencies {
+                api "org.springframework:spring-core:2.5.6"
+                constraints {
+                    implementation "commons-logging:commons-logging:1.2"
+                }
```
Comment (bigdaz) [inline subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishJavaIntegTest.groovy]: Let's add a few more dependency constraints here, to make sure they are published correctly:
- `api` vs `implementation`
- Direct dependency with no version + dependency constraint providing version. (I guess that will fail when resolving the published POM, so might need a separate test case for now).
- Constraint for `implementation` that overrides a constraint for `api` (same module, different version)
- Constraint with strict version

File: subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishJavaIntegTest.groovy [hunk 2]
```diff
@@ -201,6 +201,67 @@ class MavenPublishJavaIntegTest extends AbstractMavenPublishIntegTest {
         ]
     }
 
+    def "can publish java-library with dependency constraints"() {
```
Comment (bigdaz) [inline subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishJavaIntegTest.groovy]: I guess we should also test this for publishing to Ivy repository. (IvyPublishJavaIntegTest)

File: subprojects/docs/src/docs/design/gradle-module-metadata-specification.md [hunk 3]
```diff
@@ -101,6 +101,16 @@ Each element must be an object with the of the following values:
 
 An exclude that has a wildcard value for both `group` and `module` will exclude _all_ transitive dependencies.
 
+### `dependencyConstratins` value
```
[the remark this answers is not in the document]
Comment (jjohannes) [reply subprojects/docs/src/docs/design/gradle-module-metadata-specification.md]: Good catch! Fixed.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/GradleModuleMetadata.groovy [hunk 4]
```diff
@@ -161,6 +163,21 @@ class GradleModuleMetadata {
...
+        List<DependencyConstraint> getDependencyConstraints() {
+            if (dependencyConstraints == null) {
+                dependencyConstraints = (values.dependencyConstraints ?: []).collect {
+                    new DependencyConstraint(it.group, it.module, it.version.prefers, it.version.rejects ?: [])
+                }
+            }
+            dependencyConstraints
+        }
+
+        Variant noMoreDependencyConstraints() {
```
[the remark this answers is not in the document]
Comment (jjohannes) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/GradleModuleMetadata.groovy]: Good point. I was just mirroring what we did for dependencies thinking much about it. I changed it as you suggested.

File: subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishJavaIntegTest.groovy [hunk 5]
```diff
@@ -201,6 +201,67 @@ class MavenPublishJavaIntegTest extends AbstractMavenPublishIntegTest {
...
+        javaLibrary.parsedPom.scopes.compile.dependencies.size() == 1 //we do not publish constraints in POMs yet
+
+        and:
+        javaLibrary.parsedModuleMetadata.variant('api') {
+            dependency('org.springframework:spring-core:2.5.6') {
+                noMoreExcludes()
+                rejects()
+            }
+            noMoreDependencies()
+            noMoreDependencyConstraints()
```
[the remark this answers is not in the document]
Comment (jjohannes) [reply subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishJavaIntegTest.groovy]: Fixed.

File: subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishJavaIntegTest.groovy (hunk 1 above)
Comment (jjohannes) [reply subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishJavaIntegTest.groovy]: I am adding these as part of #3630 to also treat the consumption in the tests.

File: subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishJavaIntegTest.groovy (hunk 2 above)
Comment (jjohannes) [reply subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishJavaIntegTest.groovy]: Yeah I guess so... 
I am adding these as part of #3630 to also treat the consumption in the tests.

**Comment by jjohannes** [comment]
Thanks for the review and the very good suggestions on test coverage @bigdaz. 

I addressed the comments and merged. We now have test that show the issue with _not_ publishing constraints to maven/ivy when the version is missing. I created this follow up to address this at a later point: https://github.com/gradle/gradle/issues/3655

**Pull request merged**